homeassistant/components/ws66i/media_player.py

```diff
@@ -0,0 +1,208 @@
+        )
+
+        # Check if coordinator has data for this zone, and update values
+        data: list[ZoneStatus] = self.coordinator.data
+        if data is not None:
+            self._update_zone(data[self._zone_id_idx])
+
+    @callback
+    def _update_zone(self, zone_status: ZoneStatus) -> None:
+        """Update internal fields from a ZoneStatus."""
+        self._attr_state = STATE_ON if zone_status.power else STATE_OFF
+        self._volume = zone_status.volume
+        self._attr_is_volume_muted = zone_status.mute
+        idx = zone_status.source
+        if idx in self._ws66i_data.sources.id_name:
+            self._attr_source = self._ws66i_data.sources.id_name[idx]
+            self._attr_media_title = self._ws66i_data.sources.id_name[idx]
+        else:
+            self._attr_source = None
+            self._attr_media_title = None
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        # This will be called for each of my entities after the coordinator
+        # finishes executing _async_update_data()
+
+        # Grab the zone status that this entity represents
+        zone_status: ZoneStatus = self.coordinator.data[self._zone_id_idx]
+        self._update_zone(zone_status)
+
+        # Parent will notify HA that our fields were updated
+        super()._handle_coordinator_update()
+
+    @property
+    def volume_level(self):
+        """Volume level of the media player (0..1)."""
+        if self._volume is None:
+            return None
+        return self._volume / 38.0
+
+    def snapshot(self):
+        """Save zone's current state."""
+        self._snapshot = self._ws66i.zone_status(self._zone_id)
+
+    def restore(self):
+        """Restore saved state."""
+        if self._snapshot:
+            self._ws66i.restore_zone(self._snapshot)
+            self.schedule_update_ha_state(True)
+
+    def select_source(self, source):
+        """Set input source."""
+        if source not in self._ws66i_data.sources.name_id:
+            return
+        idx = self._ws66i_data.sources.name_id[source]
+        self._ws66i.set_source(self._zone_id, idx)
+        self.schedule_update_ha_state(True)
+
+    def turn_on(self):
+        """Turn the media player on."""
+        self._ws66i.set_power(self._zone_id, True)
+        self.schedule_update_ha_state(True)
+
+    def turn_off(self):
+        """Turn the media player off."""
+        self._ws66i.set_power(self._zone_id, False)
+        self.schedule_update_ha_state(True)
+
+    def mute_volume(self, mute):
+        """Mute (true) or unmute (false) media player."""
+        self._ws66i.set_mute(self._zone_id, mute)
+        self.schedule_update_ha_state(True)
+
+    def set_volume_level(self, volume):
+        """Set volume level, range 0..1."""
+        self._ws66i.set_volume(self._zone_id, int(volume * 38))
+        self.schedule_update_ha_state(True)
```

Comment:
```suggestion
    async def async_set_volume_level(self, volume):
        """Set volume level, range 0..1."""
        await self.hass.async_add_executor_job(self._ws66i.set_volume, self._zone_id, int(volume * 38))
        self.coordinator.async_request_refresh()
```

You can use `async_request_refresh` instead of `schedule_update_ha_state`

If you are reasonably sure the state change was successful, you could call also call `self._update_zone(zone_status)` with the new state and then call `self.async_write_ha_state()` to update the entity before doing that

Reply:
Calling self.async_write_ha_state() throws an exception in my service calls, such as def turn_off()
```
2022-03-19 21:21:05 INFO (SyncWorker_0) [homeassistant.components.ws66i.media_player] Setting power to OFF for 0
2022-03-19 21:21:05 INFO (SyncWorker_0) [homeassistant.components.ws66i.media_player] Getting state for zone 0. Power: off
2022-03-19 21:21:05 INFO (SyncWorker_0) [homeassistant.components.ws66i.media_player] Getting state for zone 0. Power: off
2022-03-19 21:21:05 INFO (SyncWorker_0) [homeassistant.components.ws66i.media_player] Getting state for zone 0. Power: off
2022-03-19 21:21:05 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [139905263015728] Non-thread-safe operation invoked on an event loop other than the current one
Traceback (most recent call last):
  File "/workspaces/core/homeassistant/components/websocket_api/commands.py", line 192, in handle_call_service
    await hass.services.async_call(
  File "/workspaces/core/homeassistant/core.py", line 1636, in async_call
    task.result()
  File "/workspaces/core/homeassistant/core.py", line 1673, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/workspaces/core/homeassistant/helpers/entity_component.py", line 204, in handle_service
    await self.hass.helpers.service.entity_service_call(
  File "/workspaces/core/homeassistant/helpers/service.py", line 668, in entity_service_call
    future.result()  # pop exception if have
  File "/workspaces/core/homeassistant/helpers/entity.py", line 931, in async_request_call
    await coro
  File "/workspaces/core/homeassistant/helpers/service.py", line 705, in _handle_entity_call
    await result
  File "/workspaces/core/homeassistant/components/media_player/__init__.py", line 649, in async_turn_off
    await self.hass.async_add_executor_job(self.turn_off)
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/workspaces/core/homeassistant/components/ws66i/media_player.py", line 188, in turn_off
    self.async_write_ha_state()
  File "/workspaces/core/homeassistant/helpers/entity.py", line 539, in async_write_ha_state
    self._async_write_ha_state()
  File "/workspaces/core/homeassistant/helpers/entity.py", line 663, in _async_write_ha_state
    self.hass.states.async_set(
  File "/workspaces/core/homeassistant/core.py", line 1373, in async_set
    self._bus.async_fire(
  File "/workspaces/core/homeassistant/core.py", line 862, in async_fire
    self._hass.async_add_hass_job(job, event)
  File "/workspaces/core/homeassistant/core.py", line 446, in async_add_hass_job
    self.loop.call_soon(hassjob.target, *args)
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 748, in call_soon
    self._check_thread()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 785, in _check_thread
    raise RuntimeError(
RuntimeError: Non-thread-safe operation invoked on an event loop other than the current one
```
In the stack trace, you will see "self.async_write_ha_state()" is called in my ws66i entity class. Is this a bug?

Reply:
You need to rename them `async_` .. and make them co-routines and wrap the call in `async_add_executor_job` to be able to use `async_write_ha_state`.  

How this works is that `async def async_set_volume_level(self, volume):` is calling `set_volume_level` under the hood
https://github.com/home-assistant/core/blob/dev/homeassistant/components/media_player/__init__.py#L665

By implementing `async_set_volume_level` we need to do the executor call ourselves.

Reply:
So coordinator updates run on the MainThread, but service calls run on a syncWorker. Calling async_write_ha_state() is not allowed from this thread. I'm going to look into how to circumvent this.

Reply:
Okay, the solution is to make all my service calls asynchronous, as mentioned in the documentation https://developers.home-assistant.io/docs/integration_fetching_data/#push-vs-poll

Reply:
Thanks for the response. I didn't refresh the page when I wrote my two other comments after your reply, so I didn't see it.